File: src/lib/graph/pagerank.ts
import type { Concept, ConceptEdge } from "@/types/concept-graph";
import Graph from "graphology";

export function calculateImportance(
  concepts: Concept[],
  edges: ConceptEdge[]
): Concept[] {
  if (concepts.length === 0) return concepts;

  const graph = new Graph({ type: "undirected" });
  for (const concept of concepts) {
    if (!graph.hasNode(concept.id)) {
      graph.addNode(concept.id);
    }
  }
  for (const edge of edges) {
    if (edge.source === edge.target) continue;
    if (!graph.hasNode(edge.source) || !graph.hasNode(edge.target)) continue;
    if (graph.hasEdge(edge.source, edge.target)) {
      const existing =
        (graph.getEdgeAttribute(
          graph.edge(edge.source, edge.target),
          "weight"
        ) as number) || 0;
      graph.setEdgeAttribute(
        graph.edge(edge.source, edge.target),
        "weight",
        existing + edge.weight
      );
    } else {
      graph.addEdge(edge.source, edge.target, { weight: edge.weight });
    }
  }

  const pagerank = computePageRank(graph, 0.85, 20);

  const prValues = Object.values(pagerank);
  const prMax = prValues.length > 0 ? Math.max(...prValues) : 0;
  const prMin = prValues.length > 0 ? Math.min(...prValues) : 0;
  const prRange = prMax - prMin || 1;

  const freqValues = concepts.map((c) => c.frequency);
  const freqMax = freqValues.length > 0 ? Math.max(...freqValues) : 1;
  const freqMin = freqValues.length > 0 ? Math.min(...freqValues) : 0;
  const freqRange = freqMax - freqMin || 1;

  return concepts.map((concept) => {
    const pr = pagerank[concept.id] ?? 0;
    const prNorm = (pr - prMin) / prRange;
    const freqNorm = (concept.frequency - freqMin) / freqRange;
    const importance = 0.7 * prNorm + 0.3 * freqNorm;
    return {
      ...concept,
      importance: Math.round(importance * 1000) / 1000,
    };
  });
}

function computePageRank(
  graph: Graph,
  damping: number,
  iterations: number
): Record<string, number> {
  const nodes = graph.nodes();
  const N = nodes.length;
  if (N === 0) return {};

  const adjacency = new Map<string, { neighbor: string; weight: number }[]>();
  const totalWeights = new Map<string, number>();

  for (const node of nodes) {
    adjacency.set(node, []);
    totalWeights.set(node, 0);
  }

  for (const edgeKey of graph.edges()) {
    const [source, target] = graph.extremities(edgeKey);
    const weight =
      (graph.getEdgeAttribute(edgeKey, "weight") as number) || 1;
    adjacency.get(source)!.push({ neighbor: target, weight });
    adjacency.get(target)!.push({ neighbor: source, weight });
    totalWeights.set(source, (totalWeights.get(source) || 0) + weight);
    totalWeights.set(target, (totalWeights.get(target) || 0) + weight);
  }

  let rank: Record<string, number> = {};
  for (const node of nodes) {
    rank[node] = 1 / N;
  }

  for (let i = 0; i < iterations; i++) {
    const newRank: Record<string, number> = {};
    let danglingSum = 0;
    for (const node of nodes) {
      if ((totalWeights.get(node) || 0) === 0) {
        danglingSum += rank[node];
      }
    }

    for (const node of nodes) {
      let sum = 0;
      const neighbors = adjacency.get(node) || [];
      for (const { neighbor, weight } of neighbors) {
        const neighborTotal = totalWeights.get(neighbor) || 0;
        if (neighborTotal > 0) {
          sum += (rank[neighbor] * weight) / neighborTotal;
        }
      }
      newRank[node] =
        (1 - damping) / N + damping * (sum + danglingSum / N);
    }
    rank = newRank;
  }

  return rank;
}

File: python-parser/app/graph/nodes/pagerank.py
"""
PageRank 重要度计算节点（对应 src/lib/graph/pagerank.ts）。

用 networkx.pagerank 替代 graphology，归一化逻辑与 TS 版本一致。
"""
from typing import Any

import networkx as nx

from app.graph.state import KGState


def calculate_importance(
    concepts: list[dict[str, Any]],
    edges: list[dict[str, Any]],
) -> list[dict[str, Any]]:
    """计算概念重要度（对应 pagerank.ts 的 calculateImportance）。

    importance = 0.7 * prNorm + 0.3 * freqNorm
    """
    if not concepts:
        return concepts

    # 构建无向图
    graph = nx.Graph()
    for concept in concepts:
        cid = concept.get("id", "")
        if cid and not graph.has_node(cid):
            graph.add_node(cid)
    for edge in edges:
        source = edge.get("source", "")
        target = edge.get("target", "")
        if source == target:
            continue
        if not graph.has_node(source) or not graph.has_node(target):
            continue
        weight = edge.get("weight", 1)
        if graph.has_edge(source, target):
            graph[source][target]["weight"] += weight
        else:
            graph.add_edge(source, target, weight=weight)

    # PageRank（damping=0.85）。某些图结构在有限迭代内不收敛，
    # 捕获 PowerIterationFailedConvergence 后用均匀分布兜底——
    # importance 计算仍可由 frequency 分量提供区分度。
    if graph.number_of_edges() > 0:
        try:
            pagerank = nx.pagerank(graph, alpha=0.85, max_iter=100, weight="weight", tol=1e-6)
        except (nx.PowerIterationFailedConvergence, Exception):
            pagerank = {n: 1.0 / graph.number_of_nodes() for n in graph}
    else:
        pagerank = {n: 1.0 / len(graph) for n in graph} if graph.number_of_nodes() > 0 else {}

    # 归一化 PageRank
    pr_values = list(pagerank.values())
    pr_max = max(pr_values) if pr_values else 0
    pr_min = min(pr_values) if pr_values else 0
    pr_range = pr_max - pr_min or 1

    # 归一化频率
    freq_values = [c.get("frequency", 0) for c in concepts]
    freq_max = max(freq_values) if freq_values else 1
    freq_min = min(freq_values) if freq_values else 0
    freq_range = freq_max - freq_min or 1

    result: list[dict[str, Any]] = []
    for concept in concepts:
        cid = concept.get("id", "")
        pr = pagerank.get(cid, 0)
        pr_norm = (pr - pr_min) / pr_range
        freq_norm = (concept.get("frequency", 0) - freq_min) / freq_range
        importance = 0.7 * pr_norm + 0.3 * freq_norm
        # round to 3 decimals（对应 TS 的 Math.round(importance * 1000) / 1000）
        result.append({**concept, "importance": round(importance, 3)})
    return result


async def calculate_importance_node(state: KGState) -> dict:
    """langgraph 节点：计算重要度。"""
    concepts = calculate_importance(state.get("concepts", []), state.get("edges", []))
    return {"concepts": concepts}
